cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchRuntimeContext.java

```diff
@@ -18,15 +18,18 @@
 
 import co.cask.cdap.api.annotation.Beta;
 import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.mapreduce.MapReduceTaskContext;
 import co.cask.cdap.etl.api.TransformContext;
 
 import java.util.Map;
 
 /**
  * Context passed to Batch Source and Sink.
+ *
+ * @param <T>  the context used to construct the BatchRuntimeContext, eg. {@link MapReduceTaskContext}
  */
 @Beta
-public interface BatchRuntimeContext extends DatasetContext, TransformContext {
+public interface BatchRuntimeContext<T> extends DatasetContext, TransformContext {
```

Comment:
good point. any suggestion? 

The only way that I can think of, is if the initialize method takes a MapReduceRuntimeContext. One way of doing that is making the BatchRuntimeContext parameter to StageLifecycle generic. However, that would mean all current plugins become mapreduce specific (because BatchSource will take the subclass of BatchRuntimeContext as a generic parameter). I wanted to keep it non-MapReduce specific, so used generics here instead. But yes, this does mean that its not very straightforward to Plugin developers.

any better ideas?